src/test/resources/com/jpinpoint/perf/lang/java/ruleset/concurrent/xml/AvoidMutableCollectionInParallelStreamForEach.xml

```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<test-data
+        xmlns="http://pmd.sourceforge.net/rule-tests"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://pmd.sourceforge.net/rule-tests http://pmd.sourceforge.net/rule-tests_1_0_0.xsd">
+    <test-code>
+        <description>violation: AvoidMutableCollectionInParallelStreamForEach</description>
+        <expected-problems>5</expected-problems>
+        <expected-linenumbers>12,32,39,57,62</expected-linenumbers>
+        <code><![CDATA[
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class AvoidMutableCollectionInParallelStreamForEach_Issue408 {
+
+    private static List<String> names = List.of("Dory", "Gill", "Bruce", "Nemo", "Darla", "Marlin", "Jacques");
+    List<String> inUppercaseField = new ArrayList<>();
+
+    public static void main(String[] args) {
+        List<String> inUppercase = new ArrayList<>();
+
+        names.parallelStream().map(String::toUpperCase) // bad 1
+                .forEach(name -> inUppercase.add(name));
+
+        System.out.println(names.size()); // 7
+        System.out.println(inUppercase.size()); // sometimes 6!
+    }
+
+    public void noBug() {
+        List<String> inUppercase = Collections.synchronizedList(new ArrayList<>());
+        names.parallelStream().map(String::toUpperCase).forEach(name -> inUppercase.add(name));
+    }
+
+    public void good() {
+        // pure functional form
+        List<String> inUppercase = names.parallelStream().map(String::toUpperCase).toList(); // java 12+
+    }
+
+    public void bad2() {
+        String[] inUppercase = new String[names.size()];
+        AtomicInteger i = new AtomicInteger(0);
+        names.parallelStream().map(String::toUpperCase) // bad 2
+                .forEach(name -> inUppercase[i.incrementAndGet()] = name);
+
+    }
+    public void bad3() {
+        Map<String, String> upperToUppercase = new HashMap<>();
+
+        names.parallelStream().map(String::toUpperCase) // bad 3
+                .forEach(name -> upperToUppercase.put(name, name));
+    }
+
+    public void good2(List inUppercase) { // don't know how implemented, could be thread-safe
```

Comment:
It is actually bad practice to modify a "foreign" list from a non private method... in practice you can also not trust that a thread safe list is provided: could recommend to make a copy and return a (new) list instead?

Reply:
Right, that would be another rule, it is suspicious. If this occurs more than once in actual code, it makes sense to create such a rule in my view